SnapEndpoints onMouseDown

SarahPappas opened the pull request.

SarahPappas committed:
Use snapEndpoints to try to snap endpoints before looking for an intersection along a path onMouseDown.

SarahPappas committed:
Create two separate snapStartPoint and snapEndPoint functions in tracingPath

msdrey commented on src/canvas/Renderer.tsx:
```diff
@@ -64,8 +64,13 @@ export class Renderer implements IRenderer {
-            const snappedPosition = this._checkPathStartIntersectionAndSplit(position, this._document.getPatternPaths());
-            this._currPath.addPoint(snappedPosition);
+            // Try to snap to other endpoints
+            const snapStartPoints = this._currPath?.snapStartPoint(this._document.getPatternPaths(), position);
```
snapStartPoint instead of snapStartPoints maybe?

msdrey commented on src/canvas/TracingPaths/TracingPath.tsx:
```diff
@@ -91,31 +90,23 @@ export abstract class TracingPath implements ITracingPath {
         }
 
-        if (updatedFirstPoint || updatedLastPoint) {
+        if (point) {
```
should this be if (updatedPoint) ?

SarahPappas replied:
yes :)

msdrey commented on src/canvas/TracingPaths/TracingPath.tsx:
```diff
@@ -68,21 +68,20 @@ export abstract class TracingPath implements ITracingPath {
 
-    snapEndToPoint = (point: Point): void => {
-        this._points[this._points.length -1] = point;
+    snapEndPoint = (paths: PatternPath[], point?: Point): boolean => {
```
It looks like the method snapEndPoint doesn't ever reassign the last point of this._points. which is kinda what the name of the method promises it would do.

SarahPappas replied:
Good catch! Thanks!

msdrey commented:
maybe if on mouse up we add the point to currpath before snapping it we could avoid having an optional parameter in snapStartPoint and snapEndPoint methods. it feels weird to call "snap my start point" on an object that has no point in it... and as a result nothing is snapped at all, we just have a true come out.

SarahPappas committed:
SnapEndPoint not updating last point in this._points

SarahPappas committed:
Update comments for snapping functions in tracingPath

SarahPappas committed:
Fix variable name

SarahPappas committed:
snapStartPoint and snapEndPoint should not take a point as a parameter

SarahPappas commented:
I agree it is weird. I changed snapStartPoint and snapEndPoint to not take points any more. I added a snapStartPointTo for our other use case in onMouseDown.
